### PR TITLE
[DI][Config] Add & use ReflectionClassResource

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #21079
| License       | MIT
| Doc PR        | -

With new changes comming to 3.3, we need a more generic reflection tracking logic than the one already managed by the autowiring subsystem.

This PR adds a new ReflectionClassResource in the Config component, and a new ContainerBuilder::getReflectionClass() method in the DI one (for fetching+tracking reflection-related info).

ReflectionClassResource tracks changes to any public or protected properties/method.

PR updated and ready, best viewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/21419/files?w=1).

changelog:

* added `ReflectionClassResource` class
* added second `$exists` constructor argument to `ClassExistenceResource` - with a special mode that prevents fatal errors from happening when some parent class is broken (logic generalized from AutowiringPass)
* made `ClassExistenceResource` also work with interfaces and traits
* added `ContainerBuilder::getReflectionClass()` for retrieving and tracking reflection class info
* deprecated `ContainerBuilder::getClassResource()`, use `ContainerBuilder::getReflectionClass()` or `ContainerBuilder::addObjectResource()` instead

### CODE REVIEW

File: src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        $tmp = sys_get_temp_dir().'/tmp.php';
+        file_put_contents($tmp, '<?php class ReflectionClassResourceTestClass {}');
+        require $tmp;
+
+        $res = new ReflectionClassResource(new \ReflectionClass('ReflectionClassResourceTestClass'));
+        $this->assertTrue($res->isFresh($now));
+
+        unlink($tmp);
+        $this->assertFalse($res->isFresh($now), '->isFresh() returns false if the resource does not exist');
+    }
```
Comment (stof) [inline src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php]: is there a way to test the handling of the hash ?

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -459,6 +424,9 @@ private function addServiceToAmbiguousType($id, $type)
         $this->ambiguousServiceTypes[$type][] = $id;
     }
 
+    /**
+     * @deprecated since version 3.3, to be removed in 4.0.
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: deprecating a private method does not make sense. Either we don't use it anymore, and we should remove it already, or we use it in deprecated method and it will become dead code when removing the other method (which is catched by IDEs and any static analysis tool, for instance SensioLabsInsight)

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 3]
```diff
@@ -279,6 +285,36 @@ public function addClassResource(\ReflectionClass $class)
     }
 
     /**
+     * Retrieves the requested reflection class and registers it for resource tracking.
+     *
+     * @param string $class
+     *
+     * @return \ReflectionClass|false
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Can we use ``\ReflectionClass|null`` instead ? It would be cleaner (PHP 7.1 has nullable types, not falsable types)

File: src/Symfony/Component/DependencyInjection/composer.json [hunk 4]
```diff
@@ -20,7 +20,7 @@
     },
     "require-dev": {
         "symfony/yaml": "~3.2",
-        "symfony/config": "~2.8|~3.0",
+        "symfony/config": "~3.3",
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/composer.json]: should we add a conflict rule to avoid issues when getting a lower version for people using standalone components ?

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 5]
```diff
@@ -56,9 +55,13 @@ public function process(ContainerBuilder $container)
      * @param \ReflectionClass $reflectionClass
      *
      * @return AutowireServiceResource
+     *
+     * @deprecated since version 3.3, to be removed in 4.0. Use ContainerBuilder::getReflectionClass() instead.
      */
     public static function createResourceForClass(\ReflectionClass $reflectionClass)
     {
+        @trigger_error('The '.__METHOD__.'() method is deprecated since version 3.3 and will be removed in 4.0. Use ContainerBuilder::getReflectionClass() instead.', E_USER_DEPRECATED);
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Not sure if it's a good idea or not. The current implementation will more efficient because some changes doesn't impact autowiring (like properties changes).

**Comment by dunglas** [comment]
👍

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: It's just a friendly reminder for the one who will cleanup the code preparing 4.0

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: would only add boilerplate (`isset(false)` is used in the code)

File: src/Symfony/Component/DependencyInjection/composer.json (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/composer.json]: good idea, added

File: src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php]: I don't see how for now :(

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: In practice, I don't think there are that more situations where the code would not-invalidate where the existing would. And I see the benefit for having only one place where this tracking logic is implemented. So I think it's in fact a desirable thing yes. Allows better aggregation of resources across passes also - and maybe faster checking (with simpler/optimized logic.)

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 3 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: Well, if this was a private API, I would be OK with this argument about simplifying the internal implementation.
But this is a public API.

Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: @nicolas-grekas I think we need to try to not use mixed return args in our public APIs (anymore). With `\ReflectionClass|null`, we could add a native PHP return type in 7.1, but `\ReflectionClass|false`, we cannot.

**Comment by nicolas-grekas** [comment]
|null done

**Comment by nicolas-grekas** [comment]
PR updated and ready, best viewed [ignoring whitespaces](https://github.com/symfony/symfony/pull/21419/files?w=1).

changelog:

* added `ReflectionClassResource` class
* added second `$exists` constructor argument to `ClassExistenceResource` - with a special mode that prevents fatal errors from happening when some parent class is broken (logic generalized from AutowiringPass)
* made `ClassExistenceResource` also work with interfaces and traits
* added `ContainerBuilder::getClassExists()` for checking and tracking class, interface or trait existence
* added `ContainerBuilder::getReflectionClass()` for retrieving and tracking reflection class info
* deprecated `ContainerBuilder::getClassResource()`, use `ContainerBuilder::getReflectionClass()` or `ContainerBuilder::addObjectResource()` instead

File: src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Tests/Resource/ReflectionClassResourceTest.php]: I split the logic to allow for testing signature mutations - tests added

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: done

**Comment by nicolas-grekas** [comment]
addObjectResource and ReflectionClassResource now track mtimes of parents, traits and interfaces.

Hash computation, file and class existence is now done lazily - ie after the resources deduplication.

And to prevent checking the hash again when it was found valid in a previous run, files are touched when needed.

**Comment by nicolas-grekas** [comment]
To be merged on top of #21452 - which holds the first commit of this PR for now.

**Comment by nicolas-grekas** [comment]
Rebased
ping @stof :)

File: src/Symfony/Component/Config/Resource/ReflectionClassResource.php [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+        return hash_final($hash);
+    }
+
+    private function generateSignature(\ReflectionClass $class)
+    {
+        yield $class->getDocComment().$class->getModifiers();
+
+        if ($class->isTrait()) {
+            yield print_r(class_uses($class->name), true);
```
Comment (stof) [inline src/Symfony/Component/Config/Resource/ReflectionClassResource.php]: shouldn't this be always included ?

File: src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php [hunk 7]
```diff
@@ -51,4 +51,20 @@ public function testIsFreshWhenClassExists()
 
         $this->assertTrue($res->isFresh(time()));
     }
+
+    public function testExistsKo()
+    {
+        spl_autoload_register(function ($class) use (&$loadedClass) { $loadedClass = $class; });
```
Comment (stof) [inline src/Symfony/Component/Config/Tests/Resource/ClassExistenceResourceTest.php]: you should unregister it at the end of the test (even when it fails)

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php [hunk 8]
```diff
@@ -30,13 +32,17 @@ public function process(ContainerBuilder $container)
...
+                continue;
             }
 
             $class = $container->getParameterBag()->resolveValue($definition->getClass());
-            if (!is_subclass_of($class, 'Symfony\\Component\\Console\\Command\\Command')) {
-                throw new \InvalidArgumentException(sprintf('The service "%s" tagged "console.command" must be a subclass of "Symfony\\Component\\Console\\Command\\Command".', $id));
+
+            if (!$r = $container->getReflectionClass($class)) {
+                throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
+            } elseif (!$r->isSubclassOf(Command::class)) {
```
Comment (stof) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: just ``if``

File: src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php [hunk 9]
```diff
@@ -30,13 +32,17 @@ public function process(ContainerBuilder $container)
             $definition = $container->getDefinition($id);
 
             if ($definition->isAbstract()) {
-                throw new \InvalidArgumentException(sprintf('The service "%s" tagged "console.command" must not be abstract.', $id));
+                continue;
```
Comment (stof) [inline src/Symfony/Component/Console/DependencyInjection/AddConsoleCommandPass.php]: why changing this ?

File: src/Symfony/Component/Config/Resource/ReflectionClassResource.php [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            }
+            if ($mtime > $timestamp) {
+                if ($this->hash !== $this->computeHash()) {
+                    return false;
+                }
+
+                // ensure the next checks won't require a hash computation again,
+                // use a local mtime cache so that the current process still sees
+                // the original mtime and validates the hash for child classes
+                @touch($file, $timestamp);
```
Comment (stof) [inline src/Symfony/Component/Config/Resource/ReflectionClassResource.php]: this might break other systems relying on the real mtime of the file because they are concerned about more things than the ones involved in the Symfony hash

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 11]
```diff
@@ -251,7 +253,32 @@ public function setResources(array $resources)
     public function addObjectResource($object)
     {
         if ($this->trackResources) {
-            $this->addClassResource(new \ReflectionClass($object));
+            if (is_object($object)) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: if it also accepts class names, the phpdoc must be updated

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 12]
```diff
@@ -263,20 +290,57 @@ public function addObjectResource($object)
...
      *
      * @return $this
+     *
+     * @deprecated since version 3.3, to be removed in 4.0. Use addObjectResource() or getReflectionClass() instead.
      */
     public function addClassResource(\ReflectionClass $class)
     {
-        if (!$this->trackResources) {
-            return $this;
+        @trigger_error('The '.__METHOD__.'() mehtod is deprecated since version 3.3 and will be removed in 4.0. Use the addObjectResource() or the getReflectionClass() method instead.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: typo

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php [hunk 13]
```diff
@@ -94,17 +95,13 @@ public function process(ContainerBuilder $container)
...
             $class = $container->getParameterBag()->resolveValue($def->getClass());
-            $interface = 'Symfony\Component\EventDispatcher\EventSubscriberInterface';
 
-            if (!is_subclass_of($class, $interface)) {
-                if (!class_exists($class, false)) {
-                    throw new InvalidArgumentException(sprintf('Class "%s" used for service "%s" cannot be found.', $class, $id));
-                }
-
-                throw new InvalidArgumentException(sprintf('Service "%s" must implement interface "%s".', $id, $interface));
+            if (!$r = $container->getReflectionClass($class)) {
```
Comment (stof) [inline src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: We actually need to invalidate the cache when the implementation of the class changes here, as it can change the list of events (btw, it means we are missing a ClassResource today)

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Config/Resource/ReflectionClassResource.php (hunk 10 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/ReflectionClassResource.php]: true, removed

File: src/Symfony/Component/Config/Resource/ReflectionClassResource.php (hunk 6 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Config/Resource/ReflectionClassResource.php]: no need: traits are not directly part of the signature - the method they provide will be seen as any other methods, thus tracked

File: src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php (hunk 13 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/EventDispatcher/DependencyInjection/RegisterListenersPass.php]: good catch, fixed

**Comment by nicolas-grekas** [comment]
@stof comments addressed

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 14]
```diff
@@ -177,16 +175,6 @@ public function dump(array $options = array())
...
 
-        foreach ($this->classResources as $r) {
-            $this->container->addClassResource($r);
-        }
-        $this->classResources = array();
-
-        foreach ($this->classResources as $r) {
-            $this->container->addClassResource($r);
-        }
-        $this->classResources = array();
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: do we actually have this duplicate code in older branches, or was it all added in 3.3 only ?

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 15]
```diff
@@ -263,20 +290,57 @@ public function addObjectResource($object)
...
-        } while ($class = $class->getParentClass());
+        } catch (\ReflectionException $e) {
+            $classReflector = false;
+        }
 
-        return $this;
+        if ($this->trackResources) {
+            if (!$classReflector) {
+                $this->addResource($resource ?: new ClassExistenceResource($class, ClassExistenceResource::EXISTS_KO));
+            } elseif (!$classReflector->isInternal()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: We can track details for internal classes, but this makes no sense: if you switch your version/extension of PHP, the container is not invalidated. Tracking here would mean checking the signature every single time. Makes no sense to me.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 14 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: 3.3 only

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php (hunk 15 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: See #21505 for internal classes

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 16]
```diff
@@ -1324,42 +1424,6 @@ public static function getServiceConditionals($value)
...
-     *
-     * @param string      $path          The file or directory path for which to check the existence
-     * @param bool|string $trackContents Whether to track contents of the given resource. If a string is passed,
-     *                                   it will be used as pattern for tracking contents of the requested directory
-     *
-     * @return bool
-     *
-     * @final
-     */
-    public function fileExists($path, $trackContents = true)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: moved as is above with other resource-related methods

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Bundle/FrameworkBundle/composer.json [hunk 17]
```diff
@@ -22,7 +22,7 @@
         "symfony/dependency-injection": "~3.3",
         "symfony/config": "~3.3",
         "symfony/event-dispatcher": "~3.3",
-        "symfony/http-foundation": "~3.1",
+        "symfony/http-foundation": "~3.3",
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: What is the reason for this change?

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: Transitivity and composer bug, this doesn't change the result, it only helps composer

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/composer.json]: thanks, I was just wondering as the code changes didn't indicate any need for this change
